# Estruturas de Repetição (Parte 1)

```
(*) Caderno preparado pelo Prof. Edson Pimentel
```

Este caderno tem por objetivo abordar conceitos relacionados a estruturas de repetição a partir de um problema.

## Base de Dados - Pacientes

Os exemplos deste caderno farão uso da base de dados pacientes.

In [ ]:
import pandas as pd 
# Base de Dados de Pacientes
#https://drive.google.com/file/d/1foZsAIE3sSoFf53ENs9QCQ29ZqcMOxac/view?usp=sharing
codigo = "1foZsAIE3sSoFf53ENs9QCQ29ZqcMOxac"
file = "https://drive.google.com/u/3/uc?id=" + codigo + "&export=download"
pac = pd.read_csv(file, sep = ";",decimal=",")
pac.head()

Ident  Idade  Altura  Peso  ...  Sistolica  Diastolica  Colesterol   IMC
0      1     58    1.80  76.1  ...        125          78         522  23.5
1      2     22    1.68  64.9  ...        107          54         127  23.0
2      3     32    1.82  80.7  ...        126          81         740  24.3
3      4     31    1.74  79.1  ...        110          68          49  26.0
4      5     28    1.72  68.7  ...        110          66         230  23.3

[5 rows x 9 columns]

## Problema 1

Considere o problema a seguir:

* Encontrar e exibir a equação da reta entre IMC (dependente) e Peso (independente).
* Exibir a predição de y (IMC) para x = mediana do Peso.

### Solução do Problema 1

A solução para este problema encontra-se a seguir. 

In [ ]:
import numpy as np
m, b = np.polyfit(x=pac['Peso'], y=pac['IMC'], deg = 1)
print("equação => y=%.2f " %m, "*x + %.2f" %b)
#---------------
mediana = pac.Peso.median()
print("mediana:", mediana)
#---------------
x = mediana
y = m*x+b
print("A predição para Peso =", x, " é IMC = %.2f" % y)

equação => y=0.23  *x + 8.35
mediana: 76.1
A predição para Peso = 76.1  é IMC = 25.57


## Problema 2

Considere o mesmo problema a acima, com o seguinte incremento:
* Exibir a predição de IMC para pesos variando entre "mediana - 1" e "mediana + 1" com variação de 0.2




### Solução do Problema 2 (tentativa)

A discussão da solução para este problema encontra-se a seguir. 

In [ ]:
x = mediana-1  # iniciando de  mediana - 1
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
x=x+0.2    # incrementando o valor de x em 0.2

A predição para Peso = 75.10  é IMC = 25.34


Note que é necessário ir incrementando o x em 0.2 e executando novamente o `mesmo trecho` de programa.

In [ ]:
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
x=x+0.2

A predição para Peso = 75.30  é IMC = 25.39


In [ ]:
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
x=x+0.2

A predição para Peso = 75.50  é IMC = 25.43


In [ ]:
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
x=x+0.2

A predição para Peso = 75.70  é IMC = 25.48


Isto precisa ser feito 11 vezes, até que x chegue em `mediana + 1 `(76.10 + 1 = 77.10)

## Solução Completa em 11 passos

In [ ]:
# completo
m, b = np.polyfit(x=pac['Peso'], y=pac['IMC'], deg = 1)
print("equação => y=%.2f " %m, "*x + %.2f" %b)
mediana = pac.Peso.median()
print("mediana:", mediana)
x = mediana-1  # iniciando de  mediana - 1
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 1
x=x+0.2    # incrementando o valor de x em 0.2 
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 2
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 3
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 4
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 5
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 6
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 7
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 8
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 9
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 10
x=x+0.2    # incrementando o valor de x em 0.2
y = m*x+b  
print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)  # -- 11
x=x+0.2    # incrementando o valor de x em 0.2

equação => y=0.23  *x + 8.35
mediana: 76.1
A predição para Peso = 75.10  é IMC = 25.34
A predição para Peso = 75.30  é IMC = 25.39
A predição para Peso = 75.50  é IMC = 25.43
A predição para Peso = 75.70  é IMC = 25.48
A predição para Peso = 75.90  é IMC = 25.52
A predição para Peso = 76.10  é IMC = 25.57
A predição para Peso = 76.30  é IMC = 25.61
A predição para Peso = 76.50  é IMC = 25.66
A predição para Peso = 76.70  é IMC = 25.70
A predição para Peso = 76.90  é IMC = 25.75
A predição para Peso = 77.10  é IMC = 25.79


### Padrão de comportamento
Note que há um padrão de comportamento (repetição) no trecho acima: 3 linhas são sempre repetidas para encontrar o próximo valor de x e y.

## Estrutura de Repetição - Enquanto

A estrutura de repetição enquanto permite especificar instruções que devem ser repetidas enquanto determinada condição for verdadeira.

A estrutura de repetição  `enquanto` `(while)` tem a seguinte estrutura:

```
while condicao:
  comandos
  comandos
  comandos
```

[Observações]:
* A "condição" é uma expressão que "compara" pelo menos dois valores com operadores relacionais (e/ou lógicos). 
> Exemplo 1: a > b
* Após a "condição" no ``while`` temos ``:`` (dois-pontos)
* Os "comandos" logo abaixo do "while" são executados se o teste de "condição" der um resultado "verdadeiro" (true)
* Os "comandos" que estão "dentro do while" são identificados pelo "deslocamento à direita" (espaços no começo da linha) em relação ao "while"
> * Estes deslocamentos são chamados de "indentação" (ou endentação)
> * Em Python o uso de indentação é obrigatório,  pois é isto que identifica o que está "dentro" ou "fora" de um "bloco".
> * Recomenda-se usar o TAB para fazer a indentação, apesar se ser possível usar também a "barra de espaço" do teclado.

### Solução do Problema 2 (com repetição)

A solução para o problema #2 usando estrutura de repetição enquanto é apresentada a seguir. Note que a solução é bem curta se comparada a solução "passo a passo" (3 linhas repetidas 11 vezes).

In [ ]:
# com repetição
x = mediana - 1
limite = mediana + 1.1
while (x <= limite):
  y = m*x+b  
  print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
  x = x + 0.2

A predição para Peso = 75.10  é IMC = 25.34
A predição para Peso = 75.30  é IMC = 25.39
A predição para Peso = 75.50  é IMC = 25.43
A predição para Peso = 75.70  é IMC = 25.48
A predição para Peso = 75.90  é IMC = 25.52
A predição para Peso = 76.10  é IMC = 25.57
A predição para Peso = 76.30  é IMC = 25.61
A predição para Peso = 76.50  é IMC = 25.66
A predição para Peso = 76.70  é IMC = 25.70
A predição para Peso = 76.90  é IMC = 25.75
A predição para Peso = 77.10  é IMC = 25.79


### Solução Completa

In [ ]:
#solução completa
import numpy as np
m, b = np.polyfit(x=pac['Peso'], y=pac['IMC'], deg = 1)
print("equação => y=%.2f " %m, "*x + %.2f" %b)
mediana = pac.Peso.median()
print("mediana:", mediana)
#------------------------------- repetição
x = mediana - 1
limite = mediana + 1.1
while (x <= limite):
  y = m*x+b  
  print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
  x = x + 0.2


equação => y=0.23  *x + 8.35
mediana: 76.1
A predição para Peso = 75.10  é IMC = 25.34
A predição para Peso = 75.30  é IMC = 25.39
A predição para Peso = 75.50  é IMC = 25.43
A predição para Peso = 75.70  é IMC = 25.48
A predição para Peso = 75.90  é IMC = 25.52
A predição para Peso = 76.10  é IMC = 25.57
A predição para Peso = 76.30  é IMC = 25.61
A predição para Peso = 76.50  é IMC = 25.66
A predição para Peso = 76.70  é IMC = 25.70
A predição para Peso = 76.90  é IMC = 25.75
A predição para Peso = 77.10  é IMC = 25.79


### Simulando o "loop infinito"

Caso a linha `x = x + 0.2` seja "comentada" a condição de "finalização" do enquanto (while) nunca será atingida e o programa entrará em **"loop infinito"**.

**[ATENÇÃO]**: Caso execute o programa abaixo com a linha "`x=x+0.2`" comentada clique em "reload"(para recarregar a página) para interromper o loop infinito.

In [ ]:
#simulando o loop infinito
# comente a linha x=x+0.2 caso queira simular o loop infinito
import numpy as np
m, b = np.polyfit(x=pac['Peso'], y=pac['IMC'], deg = 1)
print("equação => y=%.2f " %m, "*x + %.2f" %b)
mediana = pac.Peso.median()
print("mediana:", mediana)
#------------------------------- repetição
x = mediana - 1
limite = mediana + 1.1
while (x <= limite):
  y = m*x+b  
  print("A predição para Peso = %.2f" % x, " é IMC = %.2f" % y)
  x = x + 0.2   #----- linha a ser comentada para simular o loop infinito

equação => y=0.23  *x + 8.35
mediana: 76.1
A predição para Peso = 75.10  é IMC = 25.34
A predição para Peso = 75.30  é IMC = 25.39
A predição para Peso = 75.50  é IMC = 25.43
A predição para Peso = 75.70  é IMC = 25.48
A predição para Peso = 75.90  é IMC = 25.52
A predição para Peso = 76.10  é IMC = 25.57
A predição para Peso = 76.30  é IMC = 25.61
A predição para Peso = 76.50  é IMC = 25.66
A predição para Peso = 76.70  é IMC = 25.70
A predição para Peso = 76.90  é IMC = 25.75
A predição para Peso = 77.10  é IMC = 25.79


## Exercícios

### Exercício 1

Escreva um programa para:

* Encontrar e exibir a equação da reta entre Peso (dependente) e IMC(independente). Note que isto é uma inversão do exercício anterior (dependente x independente)
* Exibir o menor valor de IMC (menorIMC).
* Exibir o IMC médio (IMCmedio).
* Exibir a predição de Peso para IMC variando de "menorIMC" até "IMCmedio" com variação de 0.3

In [ ]:
#digite sua resposta aqui
import pandas as pd 
import numpy as np

#https://drive.google.com/file/d/1foZsAIE3sSoFf53ENs9QCQ29ZqcMOxac/view?usp=sharing

codigo = "1foZsAIE3sSoFf53ENs9QCQ29ZqcMOxac"
file = "https://drive.google.com/u/3/uc?id=" + codigo + "&export=download"
pac = pd.read_csv(file, sep = ";",decimal=",")
m, b = np.polyfit(x=pac['IMC'], y=pac['Peso'], deg = 1)
print("equação => y=%.2f " %m, "*x + %.2f" %b)
media = pac.IMC.mean()
menor = pac.IMC.min()
print("IMC Médio:", "%.2f" % media)
print("Menor IMC", menor)
x = menor
while( x <= media):
  y = m*x+b  
  print("A predição para IMC = %.2f" %x, " é Peso = %.2f" %y)
  x = x + 0.3

equação => y=2.58  *x + 10.92
IMC Médio: 25.98
Menor IMC 19.7
A predição para IMC = 19.70  é Peso = 61.72
A predição para IMC = 20.00  é Peso = 62.49
A predição para IMC = 20.30  é Peso = 63.27
A predição para IMC = 20.60  é Peso = 64.04
A predição para IMC = 20.90  é Peso = 64.81
A predição para IMC = 21.20  é Peso = 65.59
A predição para IMC = 21.50  é Peso = 66.36
A predição para IMC = 21.80  é Peso = 67.14
A predição para IMC = 22.10  é Peso = 67.91
A predição para IMC = 22.40  é Peso = 68.68
A predição para IMC = 22.70  é Peso = 69.46
A predição para IMC = 23.00  é Peso = 70.23
A predição para IMC = 23.30  é Peso = 71.00
A predição para IMC = 23.60  é Peso = 71.78
A predição para IMC = 23.90  é Peso = 72.55
A predição para IMC = 24.20  é Peso = 73.32
A predição para IMC = 24.50  é Peso = 74.10
A predição para IMC = 24.80  é Peso = 74.87
A predição para IMC = 25.10  é Peso = 75.65
A predição para IMC = 25.40  é Peso = 76.42
A predição para IMC = 25.70  é Peso = 77.19
